### PR TITLE
Redesign BatchedFactory (v2)

### LINKED ISSUE #317: Batch circuit executions in BatchedFactories
To speed up runs when running on, e.g., IBMQ, all scaled circuits should be generated and then sent as a single job to execute. As opposed to scaling a single circuit then single a single circuit as a job to execute.

### PR DESCRIPTION
**Pull request by rmlarose**
Fixes #317. Supersedes #364.

### CODE REVIEW

**Comment by andreamari** [comment]
Thanks @rmlarose . With the current structure the batched factory and the adaptive one have almost exactly the same methods. So what about if:

1. We go back to the old structure in which a `Factory` is adaptive by default i.e.: move the `run` method of `AdaptiveFactory` to `BaseFactory`. Remove `AdaptiveFactory` and perhaps rename `BaseFactory--->`Factory`.

2. We keep the current tree (`BaseFactory`, `BatchedFactory`, `AdaptiveFactory`) but move all the methods which are related to adaptivity (i.e. push, next, iterate, is_converged) to `AdaptiveFactory` +  re-write BatchedFactory in a more direct way without using "adaptive methods".

3. We keep the current structure of this PR. This would also be ok for me (point 1 and 2 are just minor proposals).

File: mitiq/zne/inference.py
Comment (karalekas) [inline mitiq/zne/inference.py:479]: Afaict this is identical to `AdaptiveFactory.run()`, but with `max_iterations` set to `len(self._scale_factors) + 1`. As such, if you follow @andreamari's advice (that I talk about in the previous comment above), this can be replaced by a call to `self.run` with the appropriate args, thus removing the "code smell".

File: mitiq/zne/inference.py
Comment (karalekas) [inline mitiq/zne/inference.py:467]: I believe this can just be `return self.run_batched...` as that method returns `self`.

File: mitiq/zne/inference.py
Comment (karalekas) [inline mitiq/zne/inference.py:459]: Should we somehow notify or warn the user if they are using `run_batched` but this annotation is empty? So that they know we're defaulting to non-batched?

File: mitiq/zne/inference.py
Comment (karalekas) [inline mitiq/zne/inference.py:449]: Non-blocking comment: I think we could come up a clearer adjective than "single" to describe (1).

File: mitiq/zne/inference.py
Comment (karalekas) [inline mitiq/zne/inference.py:416]: Code cleanliness comment -- looks like this method is only used by the `BatchedFactory`, and as such could be moved to this class definition.

**Review by karalekas — CHANGES_REQUESTED** [review]
Looking good! Made some suggestions in line with @andreamari's idea (1). Haven't fully considered (2), so if anyone feels strongly that that is a better alternative, I can think on it more, but so far this seems cleaner to me. As an aside, this is definitely going to conflict with #380, but I'm happy to deal with that on my end after this is merged. P.S. Don't forget to add to the changelog! :)

**Comment by rmlarose** [comment]
Thanks @andreamari and @karalekas. Just addressing the big comment now: I agree (2) is the best option - this is how it was two commits back. However, this causes the Qiskit utils tests to scale because of the noise scaling issue. (#379 is related.) To keep it manageable, how about:

1. Merge this (after addressing other comments).
1. Fix the tests/scaling method in Qiskit utils. Related to #379.
1. Implement the change suggested by @andreamari's point (2).

What do you think?

**Comment by karalekas** [comment]
Ok -- if we go with (2), do we have to rename `Factory` -> `BaseFactory`? And, regardless of the (1) vs. (2) choice, I think some of my comments are still applicable.

**Comment by rmlarose** [comment]
> Thanks @rmlarose . With the current structure the batched factory and the adaptive one have almost exactly the same methods. So what about if:
> 
> 1. We go back to the old structure in which a `Factory` is adaptive by default i.e.: move the `run` method of `AdaptiveFactory` to `BaseFactory`. Remove `AdaptiveFactory` and perhaps rename `BaseFactory--->`Factory`.
> 2. We keep the current tree (`BaseFactory`, `BatchedFactory`, `AdaptiveFactory`) but move all the methods which are related to adaptivity (i.e. push, next, iterate, is_converged) to `AdaptiveFactory` +  re-write BatchedFactory in a more direct way without using "adaptive methods".
> 3. We keep the current structure of this PR. This would also be ok for me (point 1 and 2 are just minor proposals).

Going into more detail: I'm against (1) for the following:

* This will be nearly identical to (2) because `BatchedFactory.run` will have to be overridden in the same way it is now.
* (1) makes the structure much more rigid than with `BaseFactory`, `BatchedFactory`, and `AdaptiveFactory`. The latter will make it easier to make changes in the future and is better design in my opinion. E.g., with the proposed change, `iterate` would no longer be a method of a `BatchedFactory`, only an `AdaptiveFactory`. It's easy to do this now with the `BaseFactory` - notably harder without.
* I'm pretty indifferent w.r.t `BaseFactory` vs. `Factory` (just as a name). However, I would argue for `BaseFactory` because (i) its clear that its a base class and (ii) it used to be called `Factory` but now has a different purpose.

File: mitiq/zne/inference.py
Comment (rmlarose) [reply mitiq/zne/inference.py:416]: Moved `_get_keyword_args` to `BatchedFactory` instead of `BaseFactory`.

File: mitiq/zne/inference.py
Comment (rmlarose) [reply mitiq/zne/inference.py:459]: I would say no, since:

1. If they call `run` with a non-annotated executor, `run_batched` will never be called.
1. If they call `run_batched` with a non-annotated but batched executor, we will throw a warning but shouldn't.
1. If they call `run_batched` with a non-annotated, non-batched executor, I'm willing to call that user error.

File: mitiq/zne/inference.py
Comment (rmlarose) [reply mitiq/zne/inference.py:479]: Yes it is identical. I'm proposing to not have a common run method between `BatchedFactory` and `AdaptiveFactory` (see below), so this will eventually become outdated.

File: mitiq/zne/inference.py
Comment (rmlarose) [reply mitiq/zne/inference.py:160]: See my comment below.

File: mitiq/zne/inference.py
Comment (rmlarose) [reply mitiq/zne/inference.py:449]: How about "sequential executor"?

**Comment by rmlarose** [comment]
Moved to #381.

**Pull request closed without merging**
